doctor/types.py

```diff
@@ -528,33 +531,70 @@ def get_example(cls) -> typing.Any:
 
 
 def get_value_from_schema(schema, definition: dict, key: str,
-                          definition_key: str, resolve: bool=False):
+                          definition_key: str):
     """Gets a value from a schema and definition.
 
+    If the value has references it will recursively attempt to resolve them.
+
     :param ResourceSchema schema: The resource schema.
     :param dict definition: The definition dict from the schema.
     :param str key: The key to use to get the value from the schema.
     :param str definition_key: The name of the definition.
-    :param bool resolve: If True we will attempt to resolve the definition
-        from the schema.
     :returns: The value.
     :raises TypeSystemError: If the key can't be found in the schema/definition
         or we can't resolve the definition.
     """
+    resolved_definition = definition.copy()
+    if '$ref' in resolved_definition:
+        try:
+            resolved_definition = schema.resolve(definition['$ref'])
+        except SchemaError as e:
+            raise TypeSystemError(str(e))
     try:
-        if resolve:
-            value = schema.resolve(definition['$ref'])[key]
-        else:
-            value = definition[key]
+        value = resolved_definition[key]
     except KeyError:
+        # If the key was missing and this is an array, try to resolve it
+        # from the items key.
+        if resolved_definition['type'] == 'array':
+            return [
+                get_value_from_schema(schema, resolved_definition['items'], key,
+                                      definition_key)
+            ]
+        # If the key was missing and this is an object, resolve it from it's
+        # properties.
+        elif resolved_definition['type'] == 'object':
+            value = {}
+            for prop, definition in resolved_definition['properties'].items():
+                value[prop] = get_value_from_schema(
+                    schema, definition, key, definition_key)
+            return value
+        # If the definition has a json reference, attempt to resolve it.
+        elif '$ref' in resolved_definition:
```

Comment:
It seems dangerous to have this down here. If you had a situation like:

```yaml
a:
  $ref: '#/b'
b:
  $ref: '#/c'
c:
  type: string
```

When you tried to resolve something in `a`, I think this would end up bombing on line 558 when it tries to access the `type` key after the first step resolving to `b`. Seems like we'd want to resolve any $ref before we try to access the other keys?

Reply:
Turned out the schema.resolve was already doing this recursively, but the elif was still needed because the sub-key itself might be a ref.